once_test.py

```diff
@@ -368,19 +368,98 @@ async def counting_fn2():
         self.assertEqual(await counting_fn2(), 2)
         self.assertEqual(await counting_fn2(), 2)
 
+    async def test_inspect_func(self):
+        @once.once
+        async def async_func():
+            return True
+
+        # Unfortunately these are corrupted by our @once.once.
+        # self.assertFalse(inspect.isasyncgenfunction(async_func))
+        # self.assertTrue(inspect.iscoroutinefunction(async_func))
+
+        coroutine = async_func()
+        self.assertTrue(inspect.iscoroutine(coroutine))
+        self.assertTrue(inspect.isawaitable(coroutine))
+        self.assertFalse(inspect.isasyncgen(coroutine))
+
+        # Just for cleanup.
+        await coroutine
+
+    async def test_inspect_iterator(self):
+        @once.once
+        async def async_yielding_iterator():
+            for i in range(3):
+                yield i
+
+        # Unfortunately these are corrupted by our @once.once.
+        # self.assertTrue(inspect.isasyncgenfunction(async_yielding_iterator))
+        # self.assertTrue(inspect.iscoroutinefunction(async_yielding_iterator))
+
+        coroutine = async_yielding_iterator()
+        self.assertFalse(inspect.iscoroutine(coroutine))
+        self.assertFalse(inspect.isawaitable(coroutine))
+        self.assertTrue(inspect.isasyncgen(coroutine))
+
+        # Just for cleanup.
+        async for i in coroutine:
+            pass
+
     async def test_iterator(self):
         counter = Counter()
 
-        with self.assertRaises(SyntaxError):
+        @once.once
+        async def async_yielding_iterator():
+            for i in range(3):
+                yield counter.get_incremented()
 
-            @once.once
-            async def async_yielding_iterator():
+        self.assertEqual([i async for i in async_yielding_iterator()], [1, 2, 3])
+        self.assertEqual([i async for i in async_yielding_iterator()], [1, 2, 3])
+
+    async def test_iterator_is_lazily_evaluted(self):
+        counter = Counter()
+
+        @once.once
+        async def async_yielding_iterator():
+            for i in range(3):
                 yield counter.get_incremented()
-                for i in range(3):
-                    yield i
 
-        # self.assertEqual([i async for i in async_yielding_iterator()], [1, 0, 1, 2])
-        # self.assertEqual([i async for i in async_yielding_iterator()], [1, 0, 1, 2])
+        gen_1 = async_yielding_iterator()
+        gen_2 = async_yielding_iterator()
+        gen_3 = async_yielding_iterator()
+
+        self.assertEqual(counter.value, 0)
+        self.assertEqual(await anext(gen_1), 1)
```

Comment:
It looks like `anext` was introduced in python 3.10.

Maybe we add a manaul backport?

```
if sys.version_info.minor < 10:
  async def anext(iter): ...
```

Reply:
Done!